### PR TITLE
[SimpleFSDP] Front-end Code and Unit test

### PR DESCRIPTION
**Pull request by ruisizhang123**
This PR adds front-end code of SimpleFSDP and integrated with TorchTitan's training. 

#### Composability tracker:

| Feature | Support |
| :--------: | :--------: |
|Meta Initialization| ✅ |
|Activation Checkpointing| ✅ |
|Mixed Precision Training| 🚧 |
|Tensor Parallelism| 🚧 |
|Context Parallelism| ✅ |
|Pipeline Parallelism| ✅ |
|Distributed Checkpointing|  🚧 |
|Float8 Training| ❌| 

#### Convergence on debug model

The loss curves are all benchmarked with `training.mixed_precision_param = float32` (SimpleFSDP's MPT integration is on-going) and `training.seed = 42` on 4 GPU.

- Fully Shard Mode (['dp_shard'], [4]): FSDP2 and SimpleFSDP's losses are perfectly matched

Screenshot 2025-03-20 at 10 31 22 PM

- Replicate Mode (['dp_replicate'], [4]): FSDP2 and SimpleFSDP's losses are almost the same.

Screenshot 2025-03-20 at 10 37 34 PM

**Note that the FSDP optimizations (reordering and bucketing) are not included. Thus, users won't be expected  to see the throughputs/memory numbers reported in [SimpleFSDP paper](https://arxiv.org/abs/2411.00284) only with front-end code in this pr.**

### CODE REVIEW

File: torchtitan/config_manager.py [hunk 1]
```diff
@@ -759,6 +759,11 @@ def __init__(self):
             make it available for import.
             """,
         )
+        self.parser.add_argument(
+            "--experimental.enable_simple_fsdp",
+            action="store_true",
+            help="Whether to apply SimpleFSDP to the model",
+        )
```
Comment (fegin) [inline torchtitan/config_manager.py]: Can you try `https://github.com/pytorch/torchtitan/pull/988` since this is an experimental model? Let me know what is not sufficient.

File: torchtitan/models/__init__.py [hunk 2]
```diff
@@ -7,4 +7,5 @@
 
 # Import the built-in models here so that the corresponding register_model_spec()
 # will be called.
+import torchtitan.experiments.simple_fsdp
```
Comment (fegin) [inline torchtitan/models/__init__.py]: I think we don't want the core part to depend on `torchtitan.experiments`.

File: torchtitan/train.py [hunk 3]
```diff
@@ -23,6 +23,7 @@
 )
 from torchtitan.config_manager import JobConfig
 from torchtitan.distributed import ParallelDims, utils as dist_utils
+from torchtitan.experiments.simple_fsdp import disable_simple_fsdp
```
Comment (fegin) [inline torchtitan/train.py]: ditto

**Review by fegin — COMMENTED** [review]
@ruisizhang123 Thanks again for the contribution :) We should discuss how to make the core part not depend on experiments. We may have to duplicate some methods as a first step.

File: torchtitan/experiments/simple_fsdp/README.md [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+### Enable SimpleFSDP Training
+
+Update the `CONFIG_FILE` in `run_train.sh` to `torchtitan/experiments/simple_fsdp/train_configs/debug_model.toml`
+
+### Training Convergence and Performance
+
+Training loss curve of FSDP2 and SimpleFSDP on a Llama 3.1 8B model with 8 H100 GPUs.
+
+![image](../../../assets/images/simplefsdp_loss_curves.png){width="50%"}
```
Comment (tianyu-l) [inline torchtitan/experiments/simple_fsdp/README.md]: I think we don't need to show this picture given it's in the paper.
Instead, could you fix `training.seed` and see if we can reproduce exactly the same loss from FSDP2?

It's good enough to put this in the PR summary :)

File: torchtitan/models/__init__.py (hunk 2 above)
Comment (tianyu-l) [reply torchtitan/models/__init__.py]: let's move this to `torchtitan/experiments/__init__.py`

File: torchtitan/train.py [hunk 5]
```diff
@@ -221,7 +222,7 @@ def __init__(self, job_config: JobConfig):
 
             for m in self.model_parts:
                 m.to_empty(device=init_device)
-                with torch.no_grad():
+                with disable_simple_fsdp() if job_config.experimental.enable_simple_fsdp else torch.no_grad():
```
Comment (tianyu-l) [inline torchtitan/train.py]: we don't need to change `train.py` -- we just need to specify `model.name` to be `llama3_simple_fsdp` in cmdline

File: torchtitan/experiments/simple_fsdp/train_configs/debug_model.toml [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+lr_min = 0.0
+
+[training]
+batch_size = 8
+seq_len = 2048
+max_norm = 1.0  # grad norm clipping
+steps = 10
+compile = true
+dataset = "c4_test"  # supported datasets: c4_test (2K), c4 (177M)
+mixed_precision_param = "float32"
```
Comment (tianyu-l) [inline torchtitan/experiments/simple_fsdp/train_configs/debug_model.toml]: let's remove these toml files, we should reuse `llama3_configs`
Eventually we will support mixed precision training; for now let's specify this necessary change in README.md.

File: torchtitan/experiments/simple_fsdp/parallelize_simple_fsdp.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+from torch.distributed import DeviceMesh
+from torchtitan.config_manager import JobConfig, TORCH_DTYPE_MAP
+from torchtitan.distributed import ParallelDims
+from torchtitan.models.llama.parallelize_llama import apply_ac
+from torchtitan.tools.logging import logger
+
+from .simple_fsdp import apply_simple_fsdp, MixedPrecisionPolicy
+
+
+def parallelize_simple_fsdp_llama(
```
Comment (tianyu-l) [inline torchtitan/experiments/simple_fsdp/parallelize_simple_fsdp.py]: we can still call it `parallelize_llama` -- it's in a different namespace

File: torchtitan/experiments/simple_fsdp/simple_fsdp.py [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+            output = checkpoint(
+                self.replicate_compute, x, use_reentrant=False, context_fn=fsdp_policy
+            )
+        else:
+            output = self.replicate_compute(x)
+
+        return output
+
+
+def apply_simple_fsdp(
```
Comment (tianyu-l) [inline torchtitan/experiments/simple_fsdp/simple_fsdp.py]: I'd prefer keeping the name `data_parallel` since it's not only about fsdp

File: torchtitan/experiments/simple_fsdp/README.md [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+
+### Composability Support
+
+Some of the features require the updates from PyTorch, with which we are working on providing composability support for the following features:
+
+| Feature | Support |
```
Comment (tianyu-l) [inline torchtitan/experiments/simple_fsdp/README.md:17]: let's specify
DCP as to be supported
Float8 as not supported

Let's try Context Parallel too. If it doesn't work, let's put it to "to be supported". But I somehow feel it could work.

File: torchtitan/config_manager.py (hunk 1 above)
Comment (tianyu-l) [reply torchtitan/config_manager.py]: No need to have this field.

File: torchtitan/train.py (hunk 3 above)
Comment (tianyu-l) [reply torchtitan/train.py]: we can remove this I believe.

File: torchtitan/experiments/simple_fsdp/README.md [hunk 10]
```diff
@@ -0,0 +1,39 @@
+## SimpleFSDP
+
+This folder includes an experimental frontend implementation for [SimpleFSDP: Simpler Fully Sharded Data Parallel with torch.compile](https://arxiv.org/abs/2411.00284). SimpleFSDP is a compiler-based Fully Sharded Data Parallel (FSDP) framework, which has a simple implementation for maintenance and composability, allows full computation-communication graph tracing, and brings performance enhancement via compiler backend optimizations.
+
+### Enable SimpleFSDP Training
+
+Update the `CONFIG_FILE` in `run_train.sh` to `torchtitan/experiments/simple_fsdp/train_configs/debug_model.toml`
```
Comment (tianyu-l) [inline torchtitan/experiments/simple_fsdp/README.md]: I'd imagine something like
```
CONFIG_FILE="./torchtitan/models/llama/train_configs/llama3_8b.toml" run_train.sh --model.name llama_simplefsdp --training.mixed_precision_param float32
```
would be enough

File: .github/workflows/experiments_test_8gpu.yaml
Comment (tianyu-l) [inline .github/workflows/experiments_test_8gpu.yaml:1]: Let's remove this for now. I can help set up the CI later :)

**Review by tianyu-l — COMMENTED** [review]
Overall looks amazing. Good work!
Some comments on how to organize the files.

File: torchtitan/experiments/simple_fsdp/README.md (hunk 4 above)
Comment (ruisizhang123) [reply torchtitan/experiments/simple_fsdp/README.md]: Sure, I can add this. Would running a debug model and reporting the convergence work?  I cannot run a 8B model on my side : (

File: torchtitan/train.py (hunk 3 above)
Comment (ruisizhang123) [reply torchtitan/train.py]: Thank you and Tianyu for taking time to review this pr as well! 

There might be some confusion in the naming. Basically `disable_simple_fsdp` provides the functionality of meta initialization here. It disables the parametrization in model initialization from Francisco's [PR](https://github.com/tianyu-l/torch_spmd/pull/8). 

I sync'ed with Tianyu yesterday for one bug, in which the all-gathers are not correctly gathering parameters. I later found out it was because we didn't call this `disable_simple_fsdp` and `_active_parametrization` is true in forward. But if I remove it and update corresponding changes in`simple_fsdp.py`, we won't have meta initializations lollll

I understand you don't want to put experiment in `train.py`. Do you think putting these [three line of model initialization](https://github.com/pytorch/torchtitan/blob/247f4e3aa6c65c23cf307e309eaf4a5da35dfaf0/torchtitan/train.py#L237-L239) in `self.train_spec.parallelize_fn`, instead of `train.py` would make more sense?

File: torchtitan/experiments/simple_fsdp/README.md (hunk 9 above)
Comment (ruisizhang123) [reply torchtitan/experiments/simple_fsdp/README.md:17]: yep, context parallel seems to work out of box 😁 Here is the profile trace (on debug model with ['dp_shard', 'cp'], [2, 2]). I'll update it in readme.

Screenshot 2025-03-20 at 9 32 15 PM

File: torchtitan/experiments/simple_fsdp/simple_fsdp.py (hunk 8 above)
Comment (ruisizhang123) [reply torchtitan/experiments/simple_fsdp/simple_fsdp.py]: How about `apply_data_parallel`? there is sth like `apply_fsdp`. We can match it? 🤔️

File: tests/unit_tests_experiments/simple_fsdp/__init__.py
Comment (tianyu-l) [inline tests/unit_tests_experiments/simple_fsdp/__init__.py:1]: let's put this test folder to `torchtitan/experiments/simple_fsdp/tests` -- the idea is to let the simplefsdp experiment folder be self-contained

File: tests/unit_tests_experiments/simple_fsdp/test_simple_fsdp.py [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+        fully_shard(model, mesh=self.device_mesh)
+        optim = self.optimizer(model.parameters(), lr=1e-4)
+        losses = []
+        for _ in range(epoch):
+            optim.zero_grad()
+            out = model(inputs)
+            loss = self.loss_fn(out, labels)
+            loss.backward()
+            optim.step()
+            losses.append(loss)
```
Comment (tianyu-l) [inline tests/unit_tests_experiments/simple_fsdp/test_simple_fsdp.py]: The `losses` only contain the results of the 1st forward pass, so it's not testing backward correctness.
Instead, you can run one more forward and get the loss (and append to `losses`).

File: tests/unit_tests_experiments/simple_fsdp/test_simple_fsdp.py
Comment (tianyu-l) [inline tests/unit_tests_experiments/simple_fsdp/test_simple_fsdp.py:1]: Because later we're going to have integration tests, let's rename this one to maybe `test_numerics.py` (since the folder already under simple_fsdp).

File: torchtitan/experiments/__init__.py [hunk 12]
```diff
@@ -0,0 +1,7 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+import torchtitan.experiments.simple_fsdp
```
Comment (tianyu-l) [inline torchtitan/experiments/__init__.py:7]: looks good!

File: torchtitan/train.py [hunk 13]
```diff
@@ -221,7 +222,7 @@ def __init__(self, job_config: JobConfig):
 
             for m in self.model_parts:
                 m.to_empty(device=init_device)
-                with torch.no_grad():
+                with disable_data_parallel() if job_config.model.name == "llama3_simple_fsdp" else torch.no_grad():
```
Comment (tianyu-l) [inline torchtitan/train.py]: hmm I thought about this... here's an remedy:
How about we
1. create another `model.py` under the `simple_fsdp` folder
2. define a dedicated model, e.g. called `SimpleFSDPTransformer`, which inherits the current `Transformer`
3. rewrite only the `init_weights` function, which calls `super.init_weights(...)` within the `disable_data_parallel()` context manager

It seems lighter duplicate code than copy-pasting another `train.py` just for SimpleFSDP. What do you think?

Later on we may think about abstracting out the init function into `TrainSpec`.

File: torchtitan/experiments/simple_fsdp/parallelize_llama.py [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+    parallel_dims: ParallelDims,
+    job_config: JobConfig,
+):
+    """
+    Apply activation checkpointing, torch.compile, and simplefsdp to the model.
+
+    NOTE: The passed-in model preferably should be on meta device. Otherwise,
+    the model must fit on GPU or CPU memory.
+    """
+    if job_config.activation_checkpoint.mode != "none":
```
Comment (tianyu-l) [inline torchtitan/experiments/simple_fsdp/parallelize_llama.py:59]: let's still put `apply_tp` code here and comment out with TODO, as you did for other parts involving TP

File: torchtitan/experiments/simple_fsdp/simple_fsdp.py [hunk 15]
```diff
@@ -0,0 +1,199 @@
...
+
+    def replicate_compute(self, x):
+        # data parallel runtime replicate parameters and do local compute
+        # the gradients are partial tensors that needs to perform reduction
+        # (i.e. DDP: allreduce, FSDP: reduce_scatter, HSDP: mix of both)
+
+        # NOTE: available in pytorch main
+        # if self.param_dtype is not None:
+        #     x = x.to(dtype=self.param_dtype)
+        # output = x.redistribute(placements=self.compute_placements).to_local(grad_placements=self.grad_placements)
```
Comment (tianyu-l) [inline torchtitan/experiments/simple_fsdp/simple_fsdp.py]: we can remove this comment, because the `else` branch below does this

File: torchtitan/experiments/simple_fsdp/parallelize_llama.py [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+        else:
+            dp_mesh_dim_names = ("dp_shard",)
+            fsdp_mode = "fully_shard"
+
+        mp_policy = MixedPrecisionPolicy(
+            param_dtype=TORCH_DTYPE_MAP[job_config.training.mixed_precision_param],
+            reduce_dtype=TORCH_DTYPE_MAP[job_config.training.mixed_precision_reduce],
+        )
+
+        model = apply_data_parallel(
```
Comment (tianyu-l) [inline torchtitan/experiments/simple_fsdp/parallelize_llama.py]: instead of replacing `apply_data_parallel`, we are actually replacing `fully_shard`, so I think we can just use `data_parallel`

File: torchtitan/experiments/simple_fsdp/parallelize_llama.py [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+    Apply activation checkpointing, torch.compile, and simplefsdp to the model.
+
+    NOTE: The passed-in model preferably should be on meta device. Otherwise,
+    the model must fit on GPU or CPU memory.
+    """
+    if job_config.activation_checkpoint.mode != "none":
+        apply_ac(model, job_config.activation_checkpoint)
+
+    if parallel_dims.dp_shard_enabled or parallel_dims.dp_replicate_enabled:
+        # TODO(ruisizhang123): Add support for hybrid sharding
```
Comment (tianyu-l) [inline torchtitan/experiments/simple_fsdp/parallelize_llama.py]: oh is not already supported? where would it fail?

File: torchtitan/experiments/simple_fsdp/parallelize_llama.py [hunk 18]
```diff
@@ -0,0 +1,65 @@
...
+        assert not (
+            parallel_dims.dp_shard_enabled and parallel_dims.dp_replicate_enabled
+        ), "Hybrid sharding is not supported"
+        if parallel_dims.dp_replicate_enabled:
+            dp_mesh_dim_names = ("dp_replicate",)
+            fsdp_mode = "replicate"
+        else:
+            dp_mesh_dim_names = ("dp_shard",)
+            fsdp_mode = "fully_shard"
```
Comment (tianyu-l) [inline torchtitan/experiments/simple_fsdp/parallelize_llama.py]: we should use `dp_shard_cp` mesh for fsdp wrapping, as the CP dimension can be used for parameter sharding too. Please refer to https://github.com/pytorch/torchtitan/blob/main/torchtitan/models/llama/parallelize_llama.py#L93-L96, although there it's a bit different because FSDP2 is not used for pure `dp_replicate` without `cp`
In fact, we should use `hybrid` mode for (`dp_replicate`, `dp_shard_cp`) when both are enabled.

**Review by tianyu-l — COMMENTED** [review]
I came up with a way to avoid touch `train.py`, and have a question on why HSDP won't work.

File: torchtitan/experiments/simple_fsdp/parallelize_llama.py (hunk 17 above)
Comment (ruisizhang123) [reply torchtitan/experiments/simple_fsdp/parallelize_llama.py]: yeah, we didn't test hybrid sharding before. I just did a quick verification and this works. I updated this part.

File: torchtitan/train.py (hunk 13 above)
Comment (ruisizhang123) [reply torchtitan/train.py]: yep, I think this is brilliant and makes a lot more sense. will update this.

File: tests/unit_tests_experiments/simple_fsdp/test_simple_fsdp.py (hunk 11 above)
Comment (ruisizhang123) [reply tests/unit_tests_experiments/simple_fsdp/test_simple_fsdp.py]: Not sure if I understand correctly. Here, we run the training (fwd & bwd updates) for epoch=20 times, and updating model weights. Thus, the bwd correctness is already verified? 🤔 The correctness is verified in `test_fullyshard_convergence`, `test_replicate_convergence`, and `test_hybridshard_convergence`

File: torchtitan/experiments/simple_fsdp/__init__.py [hunk 19]
```diff
@@ -0,0 +1,41 @@
...
+from .model import SimpleFSDPTransformer
+from .parallelize_llama import parallelize_llama
+from .simple_fsdp import data_parallel, disable_data_parallel
+
+__all__ = [
+    "parallelize_llama",
+    "disable_data_parallel",
+    "data_parallel",
+    "SimpleFSDPTransformer",
+]
```
Comment (tianyu-l) [inline torchtitan/experiments/simple_fsdp/__init__.py]: I think we can remove this for now, and all the unnecessary imports in the file.

**Review by tianyu-l — APPROVED** [review]
Looks fabulous! Thanks!
